### ISSUE TITLE
fix <math.h> isnan

### ISSUE THREAD

**Issue by enh** [android-ndk/ndk#113]
https://code.google.com/p/android/issues/detail?id=73725 mentions "I'm stuck building my app using NDK9d because of ... `_isnanf`".

it looks like we've always had an `isnanf` symbol, but only added `__isnanf` more recently. the NDK has a `<math.h>` whose isnan refers to __isnanf. bionic has a `<math.h>` that just says

``` c
#define isnan(x) __builtin_isnan(x)
```

unified headers fix this, but if they slip again, we should consider backporting improvements like this.

**Comment by mstorsjo** [comment]
I'm not quite sure what you intend to fix here? (Also the description has a few typos that make some sentences hard to parse.)

If you intend to make math.h for android-21 and newer keep using `isnanf` instead of `__isnanf`, that's probably a good thing to do in general, but it won't help significantly for running code built with >= android-21 headers on older versions, since the new C headers in android-21 also use lots of new functions missing on older bionic versions (rand, mkfifo, atof, etc), which according to the same bug report referenced was as intended. (Apps intended to run on older android versions strictly must be built with C library headers from an older android version as well - or are you going to revise this position?)

**Comment by TheChuckster** [comment]
@mstorsjo "it won't help significantly for running code built with >= android-21 headers on older versions"

I disagree. The idea is that it _shouldn't_ break if it doesn't _have_ to because some developers may wish to use the newer headers and optionally use the newer functions if they exist on the user's API platform at runtime.

In that sense, it will _definitely_ help. In my case, I'm stuck using NDK9d for my app because the symbols were renamed, so I can't just target API level 16 with the latest NDK, even though I'm only using the older functions. To be clear, I'm not using any of the new functions missing on older bionic versions (rand, mkfifo, atof, etc.).

One use case is that breakpad does not work on NDK9d, it only works on NDK10c+, and so we cannot use breakpad in our app because when users dynamically load my native activity library at runtime, some users' devices try to link __isnanf instead of isnanf (which is actually present) at runtime and fail because they are using older API levels:

"java.lang.UnsatisfiedLinkError: Cannot load library: reloc_library[1331]: 138 cannot locate '__isnanf'..."

It _should_ run just fine on these users' devices, except the symbol was renamed -- that's the only thing stopping it from working!

As martin.st says (in an ideal world):
"You absolutely still can build using the latest toolchain. You can use the latest NDK, but build targeting API level 16 instead of 21 (or newer), and the same single set of binaries produced should work fine on API levels both 16 and 21. There's no need to use multiple toolchains and multiple sets of binaries."

We would prefer to use the newer NDK11 for my project, but I absolutely _have_ to support users with older devices due to Android fragmentation.

See ReLinker issue:
https://github.com/KeepSafe/ReLinker/issues/15#issuecomment-198147869

This is a very simple fix that is necessary and needs to get through because of backwards compatibility breakage!

**Comment by TheChuckster** [comment]
Hope that helps you understand what @enh intends to fix here (I admit his verbiage was a bit unclear).

**Comment by mstorsjo** [comment]
> I disagree. The idea is that it shouldn't break if it doesn't have to because some developers may wish to use the newer headers and optionally use the newer functions if they exist on the user's API platform at runtime.

Yes, this part I also fully agree with. But the response I got in https://code.google.com/p/android/issues/detail?id=73725 I interpreted as that there was no intention from Android to support this; if I build using android-21 headers, they are not intending to help in any way to make such builds work, that's all up to the developer himself - the only thing they support is using the older platform headers. Now reading it in hindsight, it isn't quite explicitly said like this, but the issue was closed promptly with the resolution WorkingAsIntended.

> In that sense, it will definitely help.

Yes, it will help in your particular case. But it won't help for all the others out there, who run into issues with atof, rand, mkfifo, etc. There's lots and lots of changed functions, and this has me wondering why they suddently want to try to fix isnanf but ignoring all the other ones. Will they try to fix the other ones as well? If they only try to fix isnanf, it feels very strange and half-hearted to me.

Also, I did read all the other references you link to, and I don't see any reference to what issue you run into if you build targeting `android-16` with the latest NDK r11c (or similar). I'm genuinely interested in what this is.

To the full extent of my knowledge on the issue, this should work. If you look at `platforms/android-21/arch-arm/usr/include/math.h` in NDK r11c, you have this:

    #define isnan(x)                                        \
        ((sizeof (x) == sizeof (float)) ? __isnanf(x)       \
        : (sizeof (x) == sizeof (double)) ? isnan(x)        \
        : __isnanl(x))

Now clearly, this will try to use the new function `_isnanf` which is unavailable on older versions. Now if you instead look in `platforms/android-16/arch-arm/usr/include/math.h` in the same NDK r11c, you'll find this instead:

    #define isnan(x)                                        \
        ((sizeof (x) == sizeof (float)) ? isnanf(x)         \
        : (sizeof (x) == sizeof (double)) ? isnan(x)        \
        : __isnanl(x))

This version still refers to the old function `isnanf`, so if you build the library using the `android-16` headers in NDK r11c, I don't see how it could ever refer to the new `__isnanf` function.

There might be some other issue with using binaries built with `android-16` with NDK r11c, but I honestly don't see any reference to that in any of the links you've posted.

**Comment by DanAlbert** [comment]
@enh's message was confusing because things weren't escaped for markdown, so it dropped some things. I've edited the post to fix that. Should be more clear now :)

**Comment by mstorsjo** [comment]
> @enh's message was confusing because things weren't escaped for markdown, so it dropped some things. I've edited the post to fix that. Should be more clear now :)

Thanks!

So, is this a complete 180 degree change on this matter now, given that I reported this issue early on in the beta cycle for android L and the early NDK r10, but was closed as working as intended? Will you try to fix the C library headers for newer (>= 21) API levels to keep referring the old symbols, or will it only be fixes for some randomly picked cases? And in case you intend to fix all of them, can you confirm that is is a change in policy, since the issue was opened almost 2 years ago but promptly closed back then?

**Comment by TheChuckster** [comment]
@mstorsjo Ah, I understand what you are saying a bit better. Personally, I'm advocating for the "change in policy." Sure, it's more work, but this really does seem like unnecessary breakage to me.

"There's lots and lots of changed functions"

How many functions would you estimate? 10? 50? 100? If it's fewer than 50, there's no reason this fix shouldn't be possible.

Right, two years ago, the resolution was WorkingAsIntended with "we're working on getting weak symbols working, though, and a kind of 'compatibility library' is one of the most obvious interesting things to look at then."

It makes more sense in my mind to keep maintaining these old symbol reference as an interim fix until a new "real" fix is in place (surely, it would incentivize and hasten the development of such "real" fix, as currently, two years later, there is no such fix)...

From the referenced issue, "if you want to build with a "future" release's headers but run on an older release, you're going to be responsible for making sure that works":

Yes, but the responsibility should be in the form of "make sure you don't call new function X" not "sorry, old function X no longer works even though it's still there."

**Comment by mstorsjo** [comment]
Yes, I also would prefer this to be a change in policy. I don't have any estimate on the number of changed functions, but there's at least half a dozen, probably quite a bit more. Probably not insurmountable, but the important point I'm trying to make is that it's more than isnanf, and if this bug report only talks about that, it feels like they're missing the big issue.

> One use case (for building with the new NDK 11 targeting the older API level 16 platform) is that Google's breakpad native crash dumping library does not work on NDK9d; it only works on NDK10c+ (and believe me, the breakpad fix is not trivial -- low-level register/thread data structures in the breakpad codebase are flat out mismatched with the platform -- the devs decided to intentionally break backwards compatibility with NDK 9 at some point in time), and so we cannot use breakpad in our app because when users dynamically load my native activity library at runtime, some users' devices try to link __isnanf instead of isnanf (which is actually present) at runtime and fail because they are using older API levels:

Does breakpad fail (on newer devices, or on older devices?) if you build it targeting API level 16 with NDK r11? This is something I can relate to, that would be understandable if they touch thread internal data structures. In that case, I would say that the other major issue here is that breakpad doesn't work on the older API levels.

Or what actual issue do you run into if building (both your library and breakpad) for API level 16 with NDK r11, not with NDK r9d (which you say clearly isn't supported)? I'm really genuinely interested.

**Comment by mstorsjo** [comment]
> Yes, but the responsibility should be in the form of "make sure you don't call new function X" not "sorry, old function X no longer works even though it's still there."

Yes, also totally agree on that one - that's how I'd like it to work, even though it currently is "you need to build using the lowest API level you intend it to run on, regardless what functions you call". (I've spent a substantial amount of time the last two years helping people get this right, which is why I'm so invested in the issue.)

**Comment by TheChuckster** [comment]
"Does breakpad fail (on newer devices, or on older devices?) if you build it targeting API level 16 with NDK r11? This is something I can relate to, that would be understandable if they touch thread internal data structures. In that case, I would say that the other major issue here is that breakpad doesn't work on the older API levels.

Or what actual issue do you run into if building (both your library and breakpad) for API level 16 with NDK r11, not with NDK r9d (which you say clearly isn't supported)? I'm really genuinely interested."

Let me investigate this more today, I haven't touched this in months so I cannot remember all of the facts.

**Comment by enh** [comment]
> So, is this a complete 180 degree change on this matter now, given that I reported this issue early on in the beta cycle for android L and the early NDK r10, but was closed as working as intended? Will you try to fix the C library headers for newer (>= 21) API levels to keep referring the old symbols, or will it only be fixes for some randomly picked cases? And in case you intend to fix all of them, can you confirm that is is a change in policy, since the issue was opened almost 2 years ago but promptly closed back then?

literal answer:

i was actually saying that what you said you were doing wasn't supported --- if you're going to use dlsym, you don't need to use the newer headers at all. in fact, you probably _shouldn't_ because they'll lead you astray.

more useful answer:

bionic actually has what you want already --- its headers will give you the historical inlines if you're targeting an old API level. jmgao on the team is currently working on getting all the functions annotated with the API levels at which they were introduced. so at some point the bionic headers -- which is what we use for the platform -- will contain the historical information for the NDK too. so you'll be able to get all the latest structs, macros, and constants *regardless of API level*, with conditional visibility for the functions.

we had hoped that NDK r12 would have a preview of these "unified headers", but it's slipped to r13 at this point. you can see the progress in AOSP (in the bionic project).

long-term we plan to remove the current mess of NDK headers and go unified-only. but obviously we need to ship the new thing before we can deprecate the old thing, let alone remove it :-)

**Comment by mstorsjo** [comment]
@enh wrote:
> i was actually saying that what you said you were doing wasn't supported --- if you're going to use dlsym, you don't need to use the newer headers at all. in fact, you probably shouldn't because they'll lead you astray.

If one only need the individual functions with dlsym, then yes, you don't need newer headers. If you're going to use a complete API this way, you might need more. Concrete examples are OpenSL ES, which one would dlopen in case one supports API levels < 9, and the whole set of media/NdkMedia*.h. Currently, if one wants to use those via dlopen, one need to duplicate the enums and structs from those headers, or just drop a copy of the relevant API headers into one's own source tree, while building with a lower API level target.

I admit this is a bit of a corner case though, and definitely advanced usage. For other cases, simply building using the older API level (with the latest NDK) should be fine - hopefully we'll hear back from @TheChuckster about how that works in his case as well.

> bionic actually has what you want already --- its headers will give you the historical inlines if you're targeting an old API level. jmgao on the team is currently working on getting all the functions annotated with the API levels at which they were introduced. so at some point the bionic headers -- which is what we use for the platform -- will contain the historical information for the NDK too. so you'll be able to get all the latest structs, macros, and constants regardless of API level, with conditional visibility for the functions.
> 
> we had hoped that NDK r12 would have a preview of these "unified headers", but it's slipped to r13 at this point. you can see the progress in AOSP (in the bionic project).
> 
> long-term we plan to remove the current mess of NDK headers and go unified-only. but obviously we need to ship the new thing before we can deprecate the old thing, let alone remove it :-)

Nice, that sounds like a great thing to have!
